src/main/docs/guide/config/evaluatedExpressions.adoc

```diff
@@ -0,0 +1,376 @@
+#{ 7 >= 7 }             // true
+#{ 7 <= 8 }             // false
+
+#{ 'AbC' matches '[A-Za-z*'  }      // Compilation failure
+#{ 'AbC' matches '[A-Za-z]*'  }     // true
+#{ 'AbC' matches '[a-z]*'  }        // false
+----
+
+=== Logical Operators
+
+The following logical operators are supported: `&&` (can be aliased with `and`), `||` (can be aliased with `or`),
+`!`. Logical operations are performed in order enforced by standard operator precedence.
+You can also change evaluation order by using brackets `()`.
+
+.Logical operators examples
+[source]
+----
+#{ true && false }         // false
+#{ true and true }         // true
+
+#{ true || false }         // true
+#{ false or false }        // false
+
+#{ !false }                // true
+#{ !!true }                // true
+----
+
+=== Ternary Operator
+
+A standard ternary operator is supported to allow specifying if-then-else conditional logic in expression
+
+[source]
+----
+condition ? thenBranch : elseBranch
+----
+
+where `condition` evaluation should provide boolean value, and the complexity of `then` and `else` branches is not
+limited.
+
+.Ternary operator examples
+[source]
+----
+#{ 15 > 10 ? 'a' : 'b' }    // 'a'
+#{ 15 >= 16 ? 'a' : 'b' }   // 'b'
+----
+
+=== Type References
+
+Predefined syntax construct `T(...)` can be used to reference a class. The value inside brackets should be fully
+qualified class name (including package). The only exception is `java.lang.*` classes which can be referenced
+directly by only specifying simple class name. Primitive types can not be specified inside brackets.
+
+Type References are evaluated in different ways depending on the context.
+
+==== Simple type reference
+
+Simple type reference is resolved as `Class<?>` object.
+
+.Type reference example
+[source]
+----
+#{ T(java.lang.String) }    // String.class
+----
+
+Same rule applies if type reference is specified as method argument.
+
+==== Type check with `instanceof`
+
+Type Reference can be used as right-hand side part of `instanceof` operator
+
+.Type check example
+[source]
+----
+#{ 'abc' instanceof T(String) }  // true
+----
+
+which is equivalent to the following Java code and will be evaluated as boolean value:
+
+[source]
+----
+"abc" instanceof String
+----
+
+==== Static method invocation
+
+Type Reference can be used to invoke static method of a class
+
+.Static method invocation
+[source]
+----
+#{ T(Math).random() }
+----
+
+=== Expression Evaluation Context
+
+By default, the only methods you can invoke inside Evaluated Expressions are static methods using type references.
+
+To invoke non-static methods, you need to place `@EvaluatedExpressionContext` annotation on a class owning the
+method you want to invoke inside expression. In this case the annotated class is registered within evaluation context
+which makes its methods and properties available for referencing in evaluated expressions. Any context reference
+needs to be prefixed with `#` sign.
```

Comment:
In current implementation the user can annotate any class with `@EvaluatedExpressionContext` which will allow referencing any method or property of this class within expressions, so they do become global in this sense. Context classes can also be scoped to annotation or annotation member in the following way (forgot to mention this in docs):
```
@EvaluatedExpressionContext(CustomAnnotationLevelContext.class)
@interface CustomAnnotation {

     @EvaluatedExpressionContext(CustomAnnotationMemberContext.class)
     String customValue();
} 
```
which will limit the scope of context to respective annotation or annotation member.

The motivation behind letting users annotate any class with @EvaluatedExpressionContext was to allow something like
```
@ConfigurationProperties("config.path")
@EvaluatedExpressionContext
public class Config {
    @NotBlank
    public String cronSchedule;    
}

public class ScheduledClass {
   
    @Scheduled(cron = "#{ #cronSchedule } 
    public void runOnSchedule() {}
} 
```
another usecase is defining custom bean loading condition for usage in `@Requires`

Thinking more about it I see how that can cause issues in case of name clashes between these global context, especially if they will be defined in separate modules over which the user has no direct control. So maybe we should narrow it in some way, e.g. only allowing referencing context from the same module for classes annotated with `@EvaluatedExpressionContext` (to allow annotating only the classes that the user can modify in-place in case of clashes). And allow external modules enrich context only if annotation (or annotation member) itself has `@EvaluatedExpressionContext` on it.

Overall, the above example with `@ConfigurationProperties` is the key usecase that we would like to cover in our apps, and it seems like we can achieve it in simpler way by allowing direct reference to bean properties in expressions with special syntax construct like
```
#{ ctx[full.class.name].beanProperty } 
```
which will allow getting rid of global types. It will be more verbose, but enough for our needs, so I wouldn't mind limiting the functionality in this way as well

Reply:
I think global should be disabled by default for sure it is asking for conflicts. If needed we could add `@EvaluatedExpressionContext(global=true)` but default should be `false`

Reply:
By disabling globality you mean that context will only be available in the same module for classes that are compiled together, right? I think that introduction of flag like `global = true` is not required at the moment, would prefer more restrictive way from the start, it can always be added later if necessary

Reply:
I've added a separate annotation `@AnnotationExpressionContext` that is used as meta annotation for extending annotation and annotation member scoped context. I think that reusing `@EvaluatedExpressionContext` for user-defined classes and annotations in separate modules might be confusing as `@EvaluatedExpressionContext` had to be treated differently when class, annotation ot annotation member were annotated with it. Feels like separate annotation will make it more clear. Also mentioned it in the docs